src/csharp/Microsoft.Spark/Broadcast.cs

```diff
@@ -0,0 +1,159 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.IO;
+using System.Runtime.Serialization.Formatters.Binary;
+using Microsoft.Spark.Interop.Ipc;
+using Microsoft.Spark.Interop.Internal.Java.Util;
+using Microsoft.Spark.Interop;
+
+namespace Microsoft.Spark
+{
+    /// <summary>
+    /// A broadcast variable. Broadcast variables allow the programmer to keep a read-only variable
+    /// cached on each machine rather than shipping a copy of it with tasks. They can be used, for
+    /// example, to give every node a copy of a large input dataset in an efficient manner. Spark 
+    /// also attempts to distribute broadcast variables using efficient broadcast algorithms to 
+    /// reduce communication cost.
+    /// </summary>
+    [Serializable]
+    public sealed class Broadcast: IJvmObjectReferenceProvider
+    {
+        [NonSerialized]
+        private JvmObjectReference _jvmObject;
+        private long _bid;
+
+        internal Broadcast(SparkContext sc, object value, JvmObjectReference sparkContext)
+        {
+            var path = Path.Combine(sc._temp_dir, Path.GetRandomFileName());
+            Version version = SparkEnvironment.SparkVersion;
+
+            // Spark versions 2.3.0 and 2.3.1 create Broadcast variables using different logic
+            if (version.Major == 2 && version.Minor == 3 && (version.Build == 0 || version.Build == 1) )
+            {
+                WriteBroadcastValueToFile(sc, path, value);
+                var javaSparkContext = (JvmObjectReference)sparkContext.Jvm.CallStaticJavaMethod(
```

Comment:
Please check the coding guideline for `var` usage. Should use `JvmObjectReference`.

Reply:
I believe the coding guideline for `var` is to use it whenever the right side is an obvious type, as is true in this case. Please correct me if I'm wrong.

Reply:
Yes, you are right. var should be used with cast operator. This is the same for the rest 2 places. Sorry for the confusion.

Reply:
Yes, you are right. var should be used with cast operator. This is the same for the rest 2 places. Sorry for the confusion.